File: libs/caller-id-listener/ios/CallerIdListenerModule.swift
import ExpoModulesCore
import Foundation

public class CallerIdListenerModule: Module {
  private var server: CallerIdServer?
  private var isRunning = false
  
  public func definition() -> ModuleDefinition {
    Name("CallerIdListener")

    Events("CallerIdIncomingCall", "CallerIdListenerStatus", "CallerIdRawPacket", "CallerIdAITranscript", "CallerIdAIToolCall", "CallerIdError")

    OnCreate {
      self.server = CallerIdServer(
        onIncomingCall: { [weak self] callerNumber, callId in
          var payload: [String: Any] = [
            "phoneNumber": callerNumber,
            "timestamp": Date().timeIntervalSince1970 * 1000
          ]
          if !callId.isEmpty {
            payload["callId"] = callId
          }
          self?.sendEvent("CallerIdIncomingCall", payload)
        },
        onStatusChange: { [weak self] state, port, message in
          var payload: [String: Any] = ["state": state]
          if let port = port { payload["port"] = port }
          if let message = message { payload["message"] = message }
          
          self?.sendEvent("CallerIdListenerStatus", payload)
          
          if state == "listening" { self?.isRunning = true }
          if state == "stopped" || state == "error" { self?.isRunning = false }
        },
        onRawPacket: { [weak self] content in
          self?.sendEvent("CallerIdRawPacket", ["content": content])
        },
        onAITranscript: { [weak self] callId, role, text in
          var payload: [String: Any] = [
            "callId": callId,
            "role": role,
            "text": text
          ]
          self?.sendEvent("CallerIdAITranscript", payload)
        },
        onAIToolCall: { [weak self] callId, toolCallId, name, arguments in
          var payload: [String: Any] = [
            "callId": callId,
            "toolCallId": toolCallId,
            "name": name,
            "arguments": arguments
          ]
          self?.sendEvent("CallerIdAIToolCall", payload)
        },
        onError: { [weak self] message in
          self?.sendEvent("CallerIdError", ["message": message])
        }
      )
    }
    
    OnDestroy {
      self.server?.stop()
      self.server = nil
      self.isRunning = false
    }
    
    Function("start") { (port: Int?, sipResponses: [String]?, aiCallAssistantEnabled: Bool?, ephemeralToken: String?, fallbackNumber: String?) in
      let bindPort = UInt16(port ?? 5060)
      self.server?.start(port: bindPort, sipResponses: sipResponses ?? [], aiCallAssistantEnabled: aiCallAssistantEnabled ?? false, ephemeralToken: ephemeralToken, fallbackNumber: fallbackNumber)
    }
    
    Function("stop") {
      self.server?.stop()
      self.isRunning = false
    }
    
    Function("isRunning") {
      return self.isRunning
    }
    
    Function("sendAIToolOutput") { (callId: String, toolCallId: String, output: String) in
      self.server?.sendToolOutput(callId: callId, toolCallId: toolCallId, output: output)
    }
    
    Function("endCall") { (callId: String) in
      self.server?.endCall(callId: callId)
    }
  }
}

File: libs/caller-id-listener/android/src/main/java/expo/modules/calleridlistener/CallerIdListenerModule.kt
package expo.modules.calleridlistener

import android.os.Bundle
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class CallerIdListenerModule : Module() {

    private var server: CallerIdServer? = null
    private var isRunning = false
    private var currentPort = 5060

    override fun definition() = ModuleDefinition {
        Name("CallerIdListener")

        Events("CallerIdIncomingCall", "CallerIdListenerStatus", "CallerIdRawPacket", "CallerIdAITranscript", "CallerIdAIToolCall", "CallerIdError")

        OnCreate {
            server = CallerIdServer(
                onIncomingCall = { callerNumber, callId ->
                    val payload = Bundle().apply {
                        putString("phoneNumber", callerNumber)
                        if (callId.isNotEmpty()) {
                            putString("callId", callId)
                        }
                        putDouble("timestamp", System.currentTimeMillis().toDouble())
                    }
                    sendEvent("CallerIdIncomingCall", payload)
                },
                onStatusChange = { state, port, message ->
                    val payload = Bundle().apply {
                        putString("state", state)
                        if (port != null) putInt("port", port)
                        if (message != null) putString("message", message)
                    }
                    sendEvent("CallerIdListenerStatus", payload)
                    if (state == "listening") isRunning = true
                    if (state == "stopped" || state == "error") isRunning = false
                },
                onRawPacket = { content ->
                    val payload = Bundle().apply {
                        putString("content", content)
                    }
                    sendEvent("CallerIdRawPacket", payload)
                },
                onAITranscript = { callId, role, text ->
                    val payload = Bundle().apply {
                        putString("callId", callId)
                        putString("role", role)
                        putString("text", text)
                    }
                    sendEvent("CallerIdAITranscript", payload)
                },
                onAIToolCall = { callId, toolCallId, name, arguments ->
                    val payload = Bundle().apply {
                        putString("callId", callId)
                        putString("toolCallId", toolCallId)
                        putString("name", name)
                        putString("arguments", arguments)
                    }
                    sendEvent("CallerIdAIToolCall", payload)
                },
                onError = { message ->
                    val payload = Bundle().apply {
                        putString("message", message)
                    }
                    sendEvent("CallerIdError", payload)
                }
            )
        }

        OnDestroy {
            server?.stop()
            server = null
            isRunning = false
        }

        Function("start") { port: Int?, sipResponses: List<String>?, aiCallAssistantEnabled: Boolean?, ephemeralToken: String?, fallbackNumber: String? ->
            val bindPort = port ?: 5060
            currentPort = bindPort
            server?.start(bindPort, sipResponses ?: emptyList(), aiCallAssistantEnabled ?: false, ephemeralToken, fallbackNumber)
        }

        Function("stop") {
            server?.stop()
            isRunning = false
        }
        
        Function("sendAIToolOutput") { callId: String, toolCallId: String, output: String ->
            server?.sendToolOutput(callId, toolCallId, output)
        }

        Function("endCall") { callId: String ->
            server?.endCall(callId)
        }

        Function("isRunning") {
            return@Function isRunning
        }
    }
}
